staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/defaults.go

```diff
@@ -59,3 +59,47 @@ func SetDefaults_ServiceReference(obj *ServiceReference) {
 		obj.Port = utilpointer.Int32Ptr(443)
 	}
 }
+
+// SetDefaults_JSONSchemaProps sets the defaults for JSONSchemaProps
+func SetDefaults_JSONSchemaProps(obj *JSONSchemaProps) {
+	if obj == nil {
+		return
+	}
+	if obj.Type == "array" && obj.XListType == nil {
+		obj.XListType = utilpointer.StringPtr("atomic")
```

Comment:
won't this break non-structural schemas? those aren't allowed to use new features like kubernetes extensions.

Reply:
see schemaHasKubernetesExtensions and associated use in validation.go

Reply:
this also seems vaguely like the kind of cross-field defaulting that got us in trouble in other APIs... if I create a CRD with a schema with `type:array`, and we default an associated field that is only valid for arrays, then the creator changes the type, they're left with a defaulted field that is invalid with their new type.

Reply:
I thought the default behavior was going to be defined in the smd code, which would decide to treat unspecified lists as atomic, rather than defaulting this in the CRD schema

Reply:
> if I create a CRD with a schema with type:array, and we default an associated field that is only valid for arrays, then the creator changes the type, they're left with a defaulted field that is invalid with their new type

They are also left with a bunch of unparseable CRs?

Reply:
Not writing in the default means that if smd starts doing something smarter (I think it's a possibility) then CRs get the new behavior automatically. I can see arguments for that being both good and bad.

Reply:
I don't think we should modify the user-specified schema in CRD defaulting (especially not with logic based on other mutable fields), and should simply document how apply works for lists that do not specify their type.

Reply:
@jennybuckley I guess this is another way to have fixed the test...

Reply:
I think I would also slightly prefer to let smd do something reasonable if the user didn't directly specify.

Reply:
i can change it back to work that way